Fix param graph when mixing offset and conditional

kinow opened the pull request:
These changes close #2608 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation required.

kinow commented:
Hi @hjoliver 

While reading `param_expand.py` code, I noticed that for `m = [cat, dog]`, while expanding the graph `foo<m-1> & baz => foo<m>`, the code would start by `m=cat`. It would replace `m-1` by `-32768`, resulting in `foo_-32768 & baz => foo<m>`, and then apply a regex on it later.

The regex used is `(?:^|\s*=>).*-32768.*?(?:$|=>\s*?)`. For a given value `foo_-32768 & baz => foo_cat`, this regex matches from `foo_-32768` up to the arrow `=>` token. Figure below to simplify.

![image](https://user-images.githubusercontent.com/304786/69839803-f4004a00-12bd-11ea-8302-8f5471c10328.png)

So the returned text is ` foo_cat`. While on paper, and from our discussion in the issue linked, it ought to be `baz => foo_cat`. In this PR, I have altered the regex for `(?:^|\s*=>).*-32768([&|\s]*)?`. It is quite similar to the previous regex, changing only the last part: was `.*?(?:$|=>\s*?)` is now `([&|\s]*)?`, meaning more or less consume any spaces and `&` if any exist. Which produces:

![image](https://user-images.githubusercontent.com/304786/69839883-46416b00-12be-11ea-8140-dd4455da654f.png)

And the returned values in `param_expand` are now correct I guess, returning:

```
# a python set
{
  'baz => foo_cat',
  'foo_cat & baz => foo_dog'
}
```

Here's an example workflow I used to test it.

```
# file ~/cylc-suites/p1/suite.rc
[cylc]
    [[parameters]]
        m = cat, dog
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[dependencies]]
        [[[R1]]]
            graph = "foo<m-1> & baz => foo<m>"
```

The reference graph for Cylc 8 `master`:

```
# master
(venv) kinow@kinow-VirtualBox:~/Development/python/workspace/cylc-flow$ cylc graph --reference p1
2019-11-29T15:31:04+13:00 WARNING - deprecated graph items were automatically upgraded in 'suite definition':
2019-11-29T15:31:04+13:00 WARNING -  * (8.0.0) [scheduling][dependencies][X][graph] -> [scheduling][graph][X] - for X in:
	R1
edge "baz.1" "foo_dog.1"
edge "foo_cat.1" "foo_dog.1"
graph
node "baz.1" "baz\n1"
node "foo_cat.1" "foo_cat\n1"
node "foo_dog.1" "foo_dog\n1"
stop
```

And for this branch:

```
# this branch
(venv) kinow@kinow-VirtualBox:~/Development/python/workspace/cylc-flow$ cylc graph --reference p1
2019-11-29T15:30:54+13:00 WARNING - deprecated graph items were automatically upgraded in 'suite definition':
2019-11-29T15:30:54+13:00 WARNING -  * (8.0.0) [scheduling][dependencies][X][graph] -> [scheduling][graph][X] - for X in:
	R1
edge "baz.1" "foo_cat.1"
edge "baz.1" "foo_dog.1"
edge "foo_cat.1" "foo_dog.1"
graph
node "baz.1" "baz\n1"
node "foo_cat.1" "foo_cat\n1"
node "foo_dog.1" "foo_dog\n1"
stop
```

From what I understood, the `SuiteConfig` will call the `GraphParser` to parse the graph. The names & parameters expansion is done in `param_expand.py`, where I changed this regular expression.

But after the `GraphParser` calls the methods in `param_expand.py`'s `GraphExpander` object, it will do a further "dependency chaining". I believe this is the part that you explained to me today, where the

```
          baz => foo_cat
foo_cat & baz => foo_dog
```

Gets simplified into

```
foo_cat => foo_dog
baz => foo_cat & foo_dog
```

While it doesn't print exactly that, I think after this change the user would get the same graph & dependencies? i.e., `baz` having no dependencies, `foo_cat` having only `baz` as dependency, and `foo_dog` having `baz` and `foo_cat` as dependencies.

Added some unit tests to demonstrate the above. Tests (unit & functional) may fail, but I will deal with that later if you confirm this PR is going in the right direction :+1: 

Thanks!
Bruno

hjoliver commented:
@kinow - thanks for the fix!  From a quick look, you're right; but a proper review later.

> Gets simplified into 

I wouldn't say either one of those is simpler, they're just equivalent. And the "simplification" may have been done in my head, I wouldn't necessarily expect the graph parser to do that transformation.

kinow commented:
> > Gets simplified into
> 
> I wouldn't say either one of those is simpler, they're just equivalent. And the "simplification" may have been done in my head, I wouldn't necessarily expect the graph parser to do that transformation.

Ah! That makes sense! I thought for other graph expressions Cylc would have some sort of optimization/rewriting/etc happening somewhere for that sort of modification :smile:

kinow commented on cylc/flow/param_expand.py:
```diff
@@ -261,7 +261,7 @@ class GraphExpander(object):
     _REMOVE = -32768
     _REMOVE_REC = re.compile(
-        r'(?:^|\s*=>).*' + str(_REMOVE) + r'.*?(?:$|=>\s*?)')
+        r'(?:^|\s*=>).*' + str(_REMOVE) + r'.*?(?:$|=>\s*?|&\s*)')
```
Simpler fix. Just stop the greedy regex once it finds a) end of stream $, b) an arrow followed by spaces =>\s*, or c) ampersand symbol followed by spaces &\s*.

The last c) item covers the bug reported in the issue linked.

oliver-sanders replied:
Would these solutions handle the other forms of the problem e.g:

* `baz & foo<m-1> => foo<m>`
* `baz & foo<m-1> & pub => foo<m>`
* `bar & foo<m-1> & pub<m-1> & qux => foo<m>`

kinow replied:
@oliver-sanders not really good at parsing the graph expressions, but here's the output of the examples you provided.

For

```
params_map = {'m': ["cat", "dog"]}
templates = {'m': '_%(m)s'}
```

- `baz & foo<m-1> => foo<m>`
  * this is expanded to `{'baz & foo_cat => foo_dog', ' foo_cat'}`
- `baz & foo<m-1> & pub => foo<m>`
  * this is expanded to `{'baz & foo_cat & pub => foo_dog', 'pub => foo_cat'}`
- `bar & foo<m-1> & pub<m-1> & qux => foo<m>`
  * this is expanded to `{'bar & foo_cat & pub_cat & qux => foo_dog', 'qux => foo_cat'}`

These look OK, but can't really say I'm sure these are correct (@hjoliver do these look correct?)

oliver-sanders replied:
The expansion of the parameterised items seems correct to me but there is a bit of a discrepancy with the non-parameterised items.

E.g in example (2) we get `pub => foo_cat` but not `baz => foo_cat` similar to example (3) but different to example (1) where we just get `foo_cat`.

I'm not entirely sure how we should expand these things, here are my guesses:

* `baz & foo<m-1> => foo<m>`
  * `baz & foo_cat => foo_dog`
  * `[foo_cat]`
  * `(baz => foo_cat)`
* `baz & foo<m-1> & pub => foo<m>`
  * `baz & foo_cat & pub => foo_dog`
  * `[foo_cat]`
  * `(baz => foo_cat)`
  * `(pub => foo_cat)`
* `bar & foo<m-1> & pub<m-1> & qux => foo<m>`
  * `bar & foo_cat & pub_cat & qux => foo_dog`
  * `[foo_cat]`
  * `(bar => foo_cat)`
  * `(qux => foo_cat)`

Where:

* `compulsory => dependency` - needed for graphing
* `[optional => line]` - not needed as task is referenced elsewhere but not harmful
* `(pre-initial-type => dependency)` - not sure if we should have these or not, but if so they should be consistent

kinow commented:
Travis happy! :tada: Will push one more commit, with change log. It may fail, but it would be due to some flaky test I guess.

![image](https://user-images.githubusercontent.com/304786/79187659-fdefa780-7e70-11ea-8331-ff7e2bf5cf6c.png)

kinow commented:
(rebased)

kinow commented:
CI test failure not related to this change I think

```
 =========================== short test summary info ============================
FAILED cylc/flow/tests/network/test_client.py::TestSuiteRuntimeClient::test_serial_request
============ 1 failed, 593 passed, 1 skipped, 24 warnings in 22.25s ============
```

kinow commented:
Rebased

kinow commented:
Rebased, and moved the change log entry from 8.0a2 to 8.0a3.

kinow commented:
> Rebased, and moved the change log entry from 8.0a2 to 8.0a3.

And this time the functional tests failed (passed before rebasing). Maybe new tests caught something I had missed before :nerd_face:

kinow commented:
> > Rebased, and moved the change log entry from 8.0a2 to 8.0a3.
> 
> And this time the functional tests failed (passed before rebasing). Maybe new tests caught something I had missed before nerd_face

Kicked GH actions and now the builds passed :eyes:

hjoliver commented:
Just came back to look at this, and it doesn't seem to be working correctly (although I haven't tried to understand why). To pick one example
```
# m = cat, dog
baz & foo<m-1> & pub => foo<m> 
```
This should expand as follows:
```
baz & foo_cat & pub => foo_dog  # m = dog
baz & pub => foo_cat  # m = cat
```
In terms of individual dependencies this is:
```
# m = dog:
baz => foo_dog
foo_cat => foo_dog
pub => foo_dog

# m = cat
baz => foo_cat
pub => foo_cat
```

If I use `cylc graph` on that graph on this branch, I get all over the above except for this one:
```
baz => foo_cat  # MISSING!
```

hjoliver commented:
I'm actually not sure if the above kind of example was considered when parameters where originally implemented.  Do they make sense?  Maybe we should omit the whole line if one or more parameters goes beyond its bounds?
```
# m = cat, dog
foo<m-1> & bar => baz<m>
```
The works for `m=dog` - we get `foo_cat & bar => baz_dog`.

But for `m=cat` we could choose either:
- ignore the graph line, because `foo<m-1>` is undefined
- or ignore `foo<m-1>` (undefined) but keep `bar => baz_cat`

Maybe the former choice is more sensible?  I suspect we originally considered examples like the following, where the two choices give exactly the same result:
```
# m = 1, 2, 3
foo<m-2> => foo<m-1> => foo<m>
```
If we just ignore individual tasks that are undefined, this gives:
```
foo_m1  # m = 1
foo_m1 => foo_m2  # m = 2
foo_m1 => foo_m2 => foo_m3  # m = 3
```
Whereas if we only accept lines where all tasks are defined, we just get:
```
foo_m1 => foo_m2 => foo_m3  # m = 3  (m =1, 2 are invalid)
```
... which is the same result, because the m = 1, 2 lines in the first case are redundant with the m = 3 line.

hjoliver commented:
If doing parameterized tasks the old way, with Jinja2 loops, going beyond the list bounds is an error, so you can't loop over the whole range if there's an "index offset".  Then you can treat the edge case separately if you need it.

```
# m = 0,1,2
for m in 1, 2:
    x => foo_{m-1} => bar_{m}
y => bar_{m=0}  # m = 0 edge case if needed
```
The equivalent for parameterized tasks is:
```
x => foo<m-1> => bar<m>  # only valid for m = 1, 2, not m = 0
y => b<m=0>  # m=0 edge case if needed
```

However this still means this branch isn't working right.
```
# m = cat, dog
baz & foo<m-1> & pub => foo<m> 
```
this should generate only:
```
baz & foo_cat & pub => foo_dog
```
(and it should not generate `baz => foo_cat` or `pub => foo_cat`)




 have to treat the edge cases separately, and not go beyond the list bounds.  The equivalent for parameters, where the loops are implicit, would be: still treat the edge cases separately, and ignore the whole line if anything goes out of bounds.

hjoliver commented:
In terms of implementation, at least it should be easier to ignore whole lines that contain invalid parameterized tasks than to cut out the individual invalid tasks.  

This also gets rid of the following problem:
```
# m = cat, dog
foo => bar<m-1> => baz
```
if `m=cat` then `bar<m-1>` is undefined, so is it valid to say `foo => baz`?  NO, obviously not, if you think about the task IO files that these dependencies represent.

hjoliver commented:
(So I seem to have just proven that my own original advice on this problem was wrong, sorry @kinow :grimacing: )

oliver-sanders commented:
> Maybe the former choice is more sensible?

The issue with this is that you get inconsistent behaviour depending on the number of tasks in the parameterisation e.g:

```
# m = cat, dog
foo<m-1> & bar => baz<m>
```

Would give you:

```
foo_cat & bar => baz_dog
```

However:

```
# m = cat
foo<m-1> & bar => baz<m>
```

Would give you nothing, omitting `bar` and `baz` from the workflow entirely.

hjoliver commented:
> However:
> ```
> # m = cat
> foo<m-1> & bar => baz<m>
>```
> 
> Would give you nothing, omitting bar and baz from the workflow entirely.

Maybe that's fine - the user has written an invalid graph line given parameter has only a single value.

hjoliver commented:
I'm not sure there's a very good analogy with pre-initial logic here.  Do you think it makes sense to ignore `bar` in this expression?:
```
foo => bar<m-1> => baz
```
and get this instead?:
```
foo => baz
```

hjoliver commented:
Also, if we do what I'm suggesting, we can justify it with reference to explicit loops, where you cannot simply ignore parts of an expression where the loop index goes out of bounds.  (And parameterized tasks are supposed to be a cleaner replacement for Jinja2 loops, after all).

hjoliver commented:
> where the loop index goes out of bounds

(I mean, where the loop parameter is used as a list index, and

hjoliver commented:
(argh, accidental closure, fixed)

oliver-sanders commented:
> and get this instead?:

```
foo => baz
```

Well no that's not the correct expansion, it would be:

```
foo
baz
```

hjoliver commented:
>> I'm not sure there's a very good analogy with pre-initial logic here
> Why not, cycling and parameters are both just sequences for running tasks over with an optional offset syntax for specifying dependencies that lie on the sequence:

Actually I agree with you there.  It's more that I don't much like the pre-initial logic either. I doubt we'd have invented it in the first place if we'd had the ability to do workflow start-up properly at the time, as we do now:
```
R1 ="start => foo"
R/2/P1 = "foo[-P1] => foo"
```
That's pretty easy, clear, and correct - and no hidden pre-initial magic is needed. That said, we can at least justify the pre-initial logic somewhat on grounds that bootstrapping a cycling graph at start-up is a super-common use-case, it probably does the right thing every time(?), and we don't allow the possibly-pathological chain-breaking case (`foo => bar[-P1] => baz`).

hjoliver commented:
> No, because the out-of-bounds task would break the dependency chain so the expansion would be:
>```
>foo
>baz
>```

That's not as horrible as `foo => baz`, but it's equally unjustified isn't it?  

I suppose it doesn't matter if there parameter takes more than one value though:
```
# m = 1, 2
bar<m-1> => baz
```
It doesn't matter if this expands to only this:
```
bar_m1 => baz
```
or to this:
```
bar_m1 => baz
baz
```
because the 2nd line makes no difference when baz already appears with correct dependencies in the first line.

It only matters in the single-value case, when baz appears nowhere else in the graph.  Should this:
```
# m = 1
bar<m-1> => baz
```
be an error (or be ignored), or should it result in `baz` running with no prerequisites?:
```
baz
```

Can't say I agree with the latter choice.  The graph defines an invalid prerequisite for baz, so we should run it without any prerequisites?  (However, this is at least a pretty niche problem, I guess)

hjoliver commented:
Hmmm, actually my argument above that we should ignore a graph line if any task is out of range, is definitely wrong, because the expansion has to give the same result if each trigger is defined separately. i.e. this

```
# m = 1, 2
foo<m-1> => bar<m> => baz
```
has to be equivalent to this:
```
# m = 1, 2
foo<m-1> => bar<m> 
bar<m> => baz
```
Therefore, when interpreting the combined version, it is appropriate to simply ignore the out-of-range task.

And I suppose that does justify interpreting this:
```
# m = 1
foo<m-1> => bar
```
as this:
```
bar
```
(rather than ignoring the line entirely).

hjoliver commented:
So @kinow, having argued with myself at length until eventually ending up in @oliver-sanders's camp, I think the right thing to do on this branch is:

When going through the graph dependency pairs (`foo => bar => baz` defines [`foo => bar`, `bar => baz`]) if we encounter an out-of-range parameterized task, ignore the out of range task but not the other member of the pair:
- `foo<err> => bar` becomes `bar`
- `foo => bar<err>` becomes `foo`

(Presumably `foo<err> => bar<err>` becomes nothing though).

hjoliver commented:
(Which does mean my original advice was correct, >groan<.  I just went off-piste for a bit with the seemingly-nasty chained case).

hjoliver commented:
This is getting really tedious, but to be clear for @kinow - the same goes for the non-chained case we were originally discussing:
```
# m = [cat]  (so <m-1> is invalid)
foo & bar<m-1> & baz => qux
```
This is equivalent to:
```
foo => qux  # valid - must be included
bar<m-1> => qux  # not valid - ignore
baz => qux # valid - must be included
```
So to interpret the combined (single-line version) just ignore the out-of-bounds task: `foo & baz => qux`

kinow commented:
Rebased and moved to draft as I think it will need some modifications (reading the comments from last week).

kinow commented:
@hjoliver thanks a lot for summarizing in the last comments what's the expected result. I managed to debug it and understand what's going on. Looks like at the moment the regex captures the parameters correctly, but the regex to remove the unwanted params in a cycle (e.g. param<m-1> when m-a does not exist) removes everything from the beginning of the string.

So it computes a string like `'foo & bar_-32768 & baz => qux'`, and the regex `(?:^|\s*=>).*-32768.*?(?:$|=>\s*?|&\s*)` removes the `foo` task too. Let me see if I can change it without breaking anything else.

![image](https://user-images.githubusercontent.com/304786/89748441-609f3a00-db17-11ea-9c97-c7036399a08a.png)

Then I will go through the previous messages again to see if I missed any other case :+1:

kinow commented:
Updated the regex so that it removes only the tasks that are to be removed from an expression.

Added more test cases to the existing unit test, mainly addressing @hjoliver last comments, and also the review comment of @oliver-sanders : https://github.com/cylc/cylc-flow/pull/3452/files#diff-eead3b7f3c2d3ca8576d82a05634abc0R215

oliver-sanders commented:
Quick note, a documentation update is required in this section - https://cylc.github.io/cylc-doc/8.0a2/html/user-guide/writing-suites.html#parameter-offsets-in-the-graph

kinow commented:
Rebased

hjoliver commented:
Just tried this, still broken sorry!   The example just above:
```
[cylc]
   [[parameters]]
          m = cat   # single element, so bar<m-1> does not exist
[scheduling]
   [[graph]]
          R1 = "foo & bar<m-1> & baz => qux"
````

User `cylc graph --reference` (or running the flow) this gets expanded to this:
```
baz => qux
```

Where's `foo` gone?  The graph should get expanded to this:
```
foo => qux
baz => qux
```

(or `foo & baz => qux`)

hjoliver commented:
If I extend the parameter range by one (so `bar<m-1>` does exist, for one element):
```
[cylc]                                                                                                                                                                                                             
   [[parameters]]                                                                                                                                                                                                  
      m = cat, dog                                                                                                                                                                                                 
[scheduling]                                                                                                                                                                                                       
   [[graph]]                                                                                                                                                                                                      
      R1 = "foo & bar<m-1> & baz => qux" 
```
expands **correctly** to:
```
bar_cat => qux
baz => qux
foo => qux
```
However, this probably only seems to work correctly because the easy `m=dog` case:
```
foo & bar_cat & baz => qux
```
already correctly defines the same dependencies as the `m=cat` index-out-of-bounds edge case:
```
foo & baz => qux
```

hjoliver commented:
So: maybe the regex is removing everything up to the out-of-bounds task?
```
foo & bar<m-1> & baz => qux
```
(i.e. both `foo` and `bar` are getting nuked, when `bar<m-1>` doesn't exist - whereas only `bar` should be removed).

kinow commented:
Quickly wrote a unit test (troubleshooting some Cylc UI stuff, will look into it with more calm soon).

![image](https://user-images.githubusercontent.com/304786/90592694-4dd4e580-e23a-11ea-80f8-4f5e45c1d146.png)

And it passed :eyes:

kinow commented:
But I have successfully reproduced the error @hjoliver 

```
(venv) kinow@ranma:~/cylc-suites/cat$ cylc list cat
baz
qux
(venv) kinow@ranma:~/cylc-suites/cat$ cylc graph --reference cat
edge "baz.1" "qux.1"
graph
node "baz.1" "baz\n1"
node "qux.1" "qux\n1"
stop

```

`foo` is missing :confused: let me check what's going on…

kinow commented:
And just to confirm the expanded string actually contains `foo`, not only the expected string I used in the test.

![image](https://user-images.githubusercontent.com/304786/90592934-dd7a9400-e23a-11ea-8649-65e2708419bb.png)

hjoliver commented:
I think the difference between your test case and reality, is that in reality (if I can call it that :rofl: ) all white space gets stripped out of the graph string before it gets passed to `GraphExpander.expand()`. 

(Just running `cylc graph` through my debugger)

kinow commented:
> I think the difference between your test case and reality, is that in reality (if I can call it that rofl ) all white space gets stripped out of the graph string before it gets passed to `GraphExpander.expand()`.
> 
> (Just running `cylc graph` through my debugger)

You were fast by a couple minutes! Found the same issue with the debugger. `GraphParser` stripping spaces :angry: haha

Let me see if I can make it work with `foo&bar<m-1>&baz=>qux`.

kinow commented:
```bash
(venv) kinow@ranma:~/cylc-suites/cat$ cylc list cat
baz
foo
qux
(venv) kinow@ranma:~/cylc-suites/cat$ cylc graph --reference cat
edge "baz.1" "qux.1"
edge "foo.1" "qux.1"
graph
node "baz.1" "baz\n1"
node "foo.1" "foo\n1"
node "qux.1" "qux\n1"
stop

```

Changed regex to now tokenize the graph expression by spaces, `&`, and `|`. Then for each token it will attach the `-32768`. Then when compared with the processed graph expression, for each matched token, it will remove the token.

Then the resulting expression should be valid now, and produce the correct output :crossed_fingers: 

Added more unit tests with spaces, and with `|`'s too.

hjoliver commented:
I was going to suggest it might be easier to remove the out-of-range tasks later, after the graph is split into dependency pairs, instead of within the full graph line.   But it looks like you solved the problem anyway :+1:

hjoliver commented:
Hmmm, still one problem :grimacing:  It doesn't work in this case: 
```
foo & bar<m-1> => qux  # with m = cat only (no dog)
```
This generates no edges; it should generate `foo => qux`.

hjoliver commented:
So we need to handle:
```
REMOVE => qux  # qux
foo & REMOVE => qux  # foo => qux
REMOVE & bar => qux  # bar => qux
foo & REMOVE & bar => qux  # foo & bar => qux
```
- In the 1st case, remove `REMOVE`  (no '&'s)
- In the 2nd case, remove `& REMOVE`
- In the 3rd case, remove `REMOVE &`
- In the 4th case, remove `REMOVE &` (or `& REMOVE`? ) and leave one '&' behind.

But we can't just treat `&` as a separator, because we could have `=>` or `|` instead, on either side.

This is why I'm wondering if later pairwise removal might be easier.  If I comment out the current removal, the broken case above yields:
```
cylc graph --reference foo
edge "bar_-32768.1" "qux.1"
edge "foo.1" "qux.1"
graph
node "bar_-32768.1" "bar_-32768\n1"
node "foo.1" "foo\n1"
node "qux.1" "qux\n1"
stop
```
And we could just turn the first edge into a node `qux.1` at this point.

hjoliver commented:
(:arggh: we should have kept parameters simpler: if you want offsets like `foo<m-1>` use explicit Jinja2 loops and handle the out-of-bounds cases properly)

hjoliver commented:
Hmmm, on `foo&bar-32768=>qux` the current regex returns `foo&qux` instead of `foo=>qux`.  It works for the other cases (`bar-32768&baz=>qux` and `foo&bar-32768&baz=>qux`.

kinow commented:
> Hmmm, on `foo&bar-32768=>qux` the current regex returns `foo&qux` instead of `foo=>qux`. It works for the other cases (`bar-32768&baz=>qux` and `foo&bar-32768&baz=>qux`).

Yup, it's expanding past the last space, and including `=>` for some reason :thinking:

kinow commented:
@hjoliver still willing to test this one? I pushed my last commit, but happy to revert it if necessary. I re-wrote the regex from scratch, with verbose + comments.

```py
    _REMOVE = -32768
    # <SPACE>          : ' '
    # <AND>            : '&'
    # <OR>             : '|'
    # <TRIGGER>        : '=>'
    # <TASK_NAME_PART> : [^\s&\|] # i.e. sequence of not <AND|OR|SPACE>
    # <_REMOVE>        : '-32768'
    # <REMOVE>         : <SPACE> <TASK_NAME_PART> <_REMOVE>
    #                      <TASK_NAME_PART>? <SPACE>
    # sequence of not<AND|OR|SPACE|TRIGGER>
    _TASK_NAME_PART = r'[^\s&\|=>]'
    _REMOVE_TOKEN = rf'\s*{_TASK_NAME_PART}+{str(_REMOVE)}?' \
                    rf'{_TASK_NAME_PART}+\s*'
    _REMOVE_REC = re.compile(
        rf'''
        (                               #
          \s*=>{_REMOVE_TOKEN}(?:=>)    # <TRIGGER> <REMOVE> ?:<TRIGGER>
          |                             #
          \s*=>{_REMOVE_TOKEN}$         # <TRIGGER> <REMOVE>$
          |                             #
          ^{_REMOVE_TOKEN}=>            # ^<REMOVE> <TRIGGER>
          |                             #
          \s*[&|]+{_REMOVE_TOKEN}       # <AND|OR> <REMOVE>
          |                             #
          {_REMOVE_TOKEN}[&|]+          # <REMOVE> <AND|OR>
          |                             #
          {_REMOVE_TOKEN}               # <REMOVE>
        )                               #
        ''',
        re.VERBOSE)
```

My intention was to change as minimum as possible as I am not very familiar with how the code is used, and all the implications of changing the logic there.

But as we started finding more use cases where the old and new regexes all failed, :man_shrugging: I thought it could be worth to at least try re-writing it. If this works, I think the comments might be helpful for the next developer that needs to change that.

hjoliver commented:
Looks good, but it doesn't work in this case:
```
foo => bar<err> => baz
```
becomes `foobaz`

That chained graph is equivalent to:
```
foo => bar<err> 
bar<err> => baz
```
which should become just `foo`.  (`foo` runs, but it can't spawn an illegal `bar`; and `bar<err>` won't exist so it can't spawn `baz`)

(Ping @oliver-sanders - I think we were saying earlier that `bar<err> => baz` should become `baz` rather than nothing, but in a SoD world surely it should be nothing).

It may be unreasonable to expect a single regex to handle all these cases.  The problem is the pattern of the result is different for `&` and `=>`:
```
foo => bar<err> => baz   # should become just "foo"

foo & bar<err> & baz  # should become "foo" & "baz"
```

So I'm still leaning toward removing out-of-bounds parameterized tasks later on, once the graph is decomposed into dependency pairs...

(Sorry this one is such a pain in the :arse: @kinow !!)

kinow commented:
>So I'm still leaning toward removing out-of-bounds parameterized tasks later on, once the graph is decomposed into dependency pairs...

I'm starting to prefer this option too :grin:

kinow commented:
Phew, finally all tests passing.

@hjoliver I've gone with the approach you suggested. Instead of getting rid of the nodes out-of-range during the graph expansion step, now I'm doing that in the `GraphParser`.

My first approach was to let it build the node pairs, so using your example, I think we had:

```
foo => bar<err> => baz

None => foo
foo => bar_-32768
bar_-32768 => baz
baz => None
```

I thought I could simply skip any pairs with `-32768`. But that resulted in `foo => baz`, which you pointed would be incorrect.

Then I thought I could simply replace the error nodes by `None`. Where I would have.

```
foo => bar<err> => baz

None => foo
foo => None
None => baz
baz => None
```

I think that broke some tests and was even messier, and resulted in an exception being raised I think.

That also broke a functional test (`tests/f/validate/62-null-task-name.t`), that verifies that `t1 => => t2` results in an error being raised. But the pairs are similar to the pairs I would have if I simply replaced the error nodes by `None`. So `"foo => bar<err> => baz"` would result in an error. Also not what you suggested is the right result.

So I moved the check to just before the pairs are created. To where we have the chain of nodes.

The logic I used is the following:

- Split the line by the arrow `=>` as before
- Use the same regular expression from `GraphExpander` on each side of the expression (could be ["foo&bar", "end"], or ["foo", "end"], or ["foo", "bar", "baz"], etc)
- If one of these elements is replaced by the regular expression resulting in an empty string
  - then if it's the first token we skip it (so that "foo<err> => baz" becomes just "baz") (**is this one correct? other unit tests were passing for this…**)
  - otherwise we don't add the node to the chain and stop collecting nodes (so that "foo => bar<err> => baz" becomes just "foo", **which is what you pointed should be the correct value?**)

I didn't have to alter tests, which is a good sign (well, had to alter the graph expander as it's now not removing out-of-range nodes, so the tests now check for the existence of `errnode_-32786`). So far the PR only adds more tests, removes the graph expansion node removal regex/code, and moves the logic to graph parser.

:+1:

kinow commented on cylc/flow/graph_parser.py:
```diff
@@ -273,8 +290,29 @@ def parse_graph(self, graph_string):
+            chain = []
             # "foo => bar => baz" becomes [foo, bar, baz]
-            chain = line.split(ARROW)
+            # "foo_-32768 => bar" becomes [bar] (remove out-of-range nodes)
```
Or would this line be completely removed?

oliver-sanders replied:
If the dep is out of bounds then `[bar]` sounds right to me.

kinow commented:
Rebased, fixed conflicts. :crossed_fingers: hopefully functional & unit tests will pass again after this rebase :grimacing:

oliver-sanders commented on tests/unit/test_param_expand.py:
```diff
@@ -154,9 +154,9 @@ def test_graph_expand_offset_1(self):
-            set(["baz_i0_j0",
-                 "baz_i0_j1",
-                 "baz_i0_j2",
+            set(["bar_i-32768_j0=>baz_i0_j0",
```
Yikes!

Does this out of bounds parameter get cleared out later?

kinow replied:
It does. From what I remember of the algorithm, in the first pass it will expand the graph with these values. But later it removes them.

oliver-sanders reviewed (commented):
Thanks @kinow, heck it's been a while, sorry about that, will try to get this in quickly before we get bogged down in it again!

kinow commented:
> Thanks @kinow, heck it's been a while, sorry about that, will try to get this in quickly before we get bogged down in it again!

Thanks @oliver-sanders !

hjoliver reviewed (changes requested):
The new logic doesn't quite work sorry!  (although it should be much easier to modify, to make it work, I hope).

This:
```
foo => bar => baz
```
is equivalent to this:
```
foo => bar
bar => baz
```
so they must yield the same result after parsing.

In the parameter case though:
```
# m = cat
foo => bar<m-1> => baz
```
the split (two line) version gives a different result (via `cylc graph`) because your logic treats the first token in the chain as special.

kinow commented:
So close!!!

hjoliver commented:
I think this shows we can't treat out-of-bounds parameters like pre-initial cycle dependence, because of the fact that out-of-bounds parameters can appear in on the right side of a trigger arrow.

So:
```
# m = cat
foo => bar<m-1> => baz
```
This means:
```
foo => ERR => baz
```
This should yield only `foo`.  :heavy_check_mark: 

```
# m = cat
foo => ERR
ERR => baz
```
This should also yield only `foo` BUT currently it yields `foo` and `baz` :fearful: 

IMO what I'm arguing makes sense in spite of @oliver-sanders pre-initial-logic based argument above because:
- 1) consistency between the single and multi-line versions of the same expression
- 2) pre-initial dependence is different because it can only ever be on the left side of a chain
- 3) `ERR => baz` says "baz triggers off of an illegal/non-existent task" - clearly the right answer to that is "baz should not trigger" (it is NOT "baz should trigger immediately as if it had no prerequisites")

hjoliver commented:
More thoughts on the pre-initial analogy:
```
# initial cycle point = 1
P1 = foo[-P1] => bar
```
Here `foo.0` is technically an illegal/non-existent task BUT we ignore `foo.0` to get just `bar.1` as a convenience to allow users to boot-strap into an infinite cycle without having to handle the initial cycle properly. i.e. we're are assuming the user really means this:
```
R1 = bar
R/2/P1 = foo[-P1] => bar
```

In the parameter case, the same logic arguably makes sense here
```
# m = 1,2,3,...
foo<m-1> => bar<m>
```
but does it make sense here:?
```
# m = 1,2,3 ...
foo<m-1> => bar
```
(This is another difference between parameters and pre-initial cycles: not all tasks are parameterized, but all tasks do have cycle point "parameter").

hjoliver commented:
The good news is, maybe it only matters for the case of a single out-of-bounds parameter. This should expand to nothing, not `bar`:
```
# m = cat
foo<m-=1> => bar
```
but this could expand to`bar` (m=cat) as well as  `foo<cat> => bar` (m=dog) because the total result is the same (bar depends on foo_cat).

hjoliver commented:
Upshot: this issue is a nightmare :grimacing:  However, it seems to me that we have to go with what I've suggested above because of the fact that parameterized tasks can appear on the right side of a trigger, and we need consistency between chained and list-of-pairs graph strings.  Would you agree @oliver-sanders ? (before @kinow bashes his head on this wall any further).

oliver-sanders commented:
Oh dammit, this one is the gift that just keeps giving.

I'm going to have to go and work through some examples to suss out the implications.

hjoliver commented:
@kinow - I meant to make a PR to your branch, but I accidentally pushed the commit directly to it. Doh!  (not a disaster, as it can be backed out if necessary, of course).  Anyhow, it simplifies your new logic to make it do what I think is correct, which is: in a chain of dependencies stop at an out-of-bounds task (whether it is the first on in the chain or not).

```
# m = cat
foo => bar => baz<m-1>   # becomes foo => bar
foo => bar<m-1> => baz   # becomes foo
foo<m-1> => bar => baz   # becomes nothing
```

If it passes all the test, we'll just wait to see if @oliver-sanders can convince himself that this is the correct way to go.

Note:
- this logic is easy to explain: an out-of-bounds task is illegal, so it can't trigger other tasks downstream of it
- if users really want the alternative logic where the out of bounds task appears at the start of the chain, they can still do it:
```
foo<m-1> => bar => baz
bar => bar
```

hjoliver commented:
(repushed with style fix)

kinow commented:
Feel free to update this branch/PR as necessary :) Thanks @hjoliver !

hjoliver commented:
Tests all pass :tada:

oliver-sanders commented:
```bash
## the parameters
non =
one = 1
two = 1, 2

$> the outputs

## an empty parameter yields no tasks
$> Nothing
<non>

## a non-empty parameter does yield tasks
> 1
<one>

## a dependency on an empty parameter under SoD is like a dependency on a task
## that is not in the graph, <one> will never run so we might as well ignore it here???
$> Nothing
<non> => <one>

## ok but what about the other way around if:
##     a => b
## is equivalent to
##     a
##     a => b
## then the same should be true for parameters?
##    <one>
##    <one> => <non>
## (I'm going to go with yes):
$> 1
<one> => <non>

## ok, now for offsets...
## a parameter expands simply
$> 1 => a
$> 2 => a
<two> => a

## but how does a parameter offset like this expand?
## Is is equivalent to:
##     <two>
##     <two - 1> => a
## or:
##     <two - 1>
##     <two - 1> => a
## (I'm going to go with <two - 1> == [1])
$> 1 => a
<two - 1> => a

## if you stick another task at the beginning of the chain, the offset
## should still behave the same, but should the head of the chain expand
##    <two>
##    <two> => <two -1> => a
$> 1 # ???
$> 2 =1 => a
<two> => <two - 1> => a

## If so
$> 1
<one> => <one - 1> => a
```

1. Ignore everything in the chain after the first out-of-bounds parameter.
2. The first item in the chain is always added irrespective of what follows.

hjoliver commented:
@oliver-sanders - in your example above, does "non" mean:
- 1) an empty parameter list (`m = # nothing` with `foo<m>`) 
      - this fails validation
- 2) or a non-parameterized task (`foo`)
     - this doesn't fit with your statement that "## an empty parameter yields no tasks"
- 3) or an out-of-bounds parameter?
    - this doesn't seem to fit either, since you're addressing that with the "one" and "two" parameters

It seems to me you mean (i); however that fails validation so we don't need to worry about it (which I think is sensible).

So we only need to address:
- non-parameterized tasks 
- parameterized tasks with valid parameter values 
- parameterized tasks with invalid or out-of-bounds parameter values

Maybe I didn't clearly state my rule above in one place:
- **remove tasks with invalid parameter values, and all tasks downstream of them**

I think that corresponds to your rule 1
> Ignore everything in the chain after the first out-of-bounds parameter?

but I'm not sure about your rule 2 ??
> The first item in the chain is always added irrespective of what follows?

I would say don't add the first item either if it has an invalid parameter value.  

E.g. for 
```
# m = 1,2
foo<m-1> => bar<m> => baz
```
This becomes:
```
foo_1 => bar_2 => baz   # m = 2
# (nothing)                       # m = 1
```
When `m=1` we don't retain `bar_1 => baz` because the graph says `bar_1` is supposed to trigger off of an illegal task.  As I said in previous comments, if we did try to allow that (which is somewhat analogous with our handling of pre-initial dependence) it seems problematic when the out-of-bounds parameter is not at the front of the chain:
```
# m = 1
foo => bar<m-1> => baz
```
My rule says this should be interpreted as `foo`.  Surely `foo => baz` is not reasonable ???   Maybe interpreting it as `foo & baz` would be acceptable, but a convincing use case would be nice because it seems less reasonable to me than saying `baz` should not trigger when `<m-1>` is invalid.

oliver-sanders commented:
> @oliver-sanders - in your example above, does "non" mean:

An empty parameter list (m = # nothing with foo<m>) 

> The first item in the chain is always added irrespective of what follows?

What I'm asking is should this:

```
# m = 1,2
foo<m> => bar<m-1> => baz
```

Result in this:

```
foo_1 => bar_2 => baz   # m = 1
foo_2                   # m = 2
```

Or this:

```
foo_1 => bar_2 => baz   # m = 1
# (nothing)             # m = 2
```

Based on the logic that this:

```
foo => bar
```

Is equivalent to this:

```
foo
foo => bar
```

I'm inclined to go with the first example.

hjoliver commented:
Ah, I think your example got munged.

Presumably what you mean is, should this:
```
# m = 1,2
foo<m> => bar<m-1> => baz
```
Result in this:
```
# Option 1
foo_2 => bar_1 => baz   # m = 2
foo_1                   # m = 1
```
Or this:
```
# Option 2
foo_2 => bar_1 => baz   # m = 2
# (nothing)             # m = 1
```

If that's what you meant, then we're in agreement that Option 1 is correct. And that's what the current implementation does:
```
$ cylc graph --reference flow.cylc

edge "bar_m1.1" "baz.1"       # <----
edge "foo_m2.1" "bar_m1.1"    # <----
graph
node "bar_m1.1" "bar_m1\n1"
node "baz.1" "baz\n1"
node "foo_m1.1" "foo_m1\n1"   # <----
node "foo_m2.1" "foo_m2\n1"
stop
```

hjoliver commented:
@oliver-sanders - requesting re-approval from you (it looks like you are happy with it, based on previous discussion, but just in case).

oliver-sanders commented:
Yep looks like we are in agreement, good, bye bye to one of our oldest PRs, sorry @kinow!

oliver-sanders commented:
Quick lets just get this in before we change our minds again!
